csharp/LogicalColumnWriter.cs

```diff
@@ -106,136 +108,143 @@ public override void WriteBatch(ReadOnlySpan<TElement> values)
             // Handle arrays separately
             if (typeof(TElement) != typeof(byte[]) && typeof(TElement).IsArray)
             {
-                if (NestingDepth == 0)
-                {
-                    WriteArrayFinalLevel(values, 0, 0, ColumnDescriptor.MaxDefinitionLevel, converter, NullDefinitionLevels[0]);
-                }
-                else
-                {
-                    WriteArrayIntermediateLevel(values, 0, NestingDepth, 0, NullDefinitionLevels, ColumnDescriptor.MaxDefinitionLevel, converter);
-                }
+                WriteArray(values, ColumnDescriptor.SchemaNode, converter);
             }
             else
             {
                 WriteBatchSimple(values, converter as LogicalWrite<TElement, TPhysical>.Converter);
             }
         }
 
-        private void WriteArrayIntermediateLevel(
-            ReadOnlySpan<TElement> values,
-            short repetitionLevel, short maxRepetitionLevel, short leafFirstRepLevel, 
-            short[] nullDefinitionLevels, short leafDefinitionLevel, 
-            LogicalWrite<TLogical, TPhysical>.Converter converter)
+        private static List<Node> GetSchemaNode(Node node)
```

Comment:
Duplicated with reader, belongs in base class.

Also, as discussed, done and checked at ctor time so it does not need to be called at each read.